js/views/install_view.js

```diff
@@ -14,145 +14,148 @@
+                    errorMessage = i18n('installTooManyDevices');
+                }
+                else if (this.error.name === 'HTTPError' && this.error.code == 411) {
+                    errorMessage = i18n('installConnectionFailed');
+                }
+                else if (this.error.message === 'websocket closed') {
+                    errorMessage = i18n('installConnectionFailed');
+                }
 
-            if (Whisper.Registration.everDone()) {
-                this.selectStep(Steps.SCAN_QR_CODE);
-                this.hideDots();
+                return {
+                    error: true,
+                    errorHeader: 'Something went wrong!',
+                    errorMessage,
+                    errorButton: 'Try again',
+                };
             }
+
+            return {
+                step3: this.step === Steps.SCAN_QR_CODE,
+                linkYourPhone: i18n('linkYourPhone'),
+                signalSettings: i18n('signalSettings'),
+                linkedDevices: i18n('linkedDevices'),
+                androidFinalStep: i18n('plusButton'),
+                appleFinalStep: i18n('linkNewDevice'),
+
+                step4: this.step === Steps.ENTER_NAME,
+                chooseName: i18n('chooseDeviceName'),
+                finishLinkingPhoneButton: i18n('finishLinkingPhone'),
+
+                step5: this.step === Steps.PROGRESS_BAR,
+                syncing: i18n('initialSync'),
+            };
+        },
+        selectStep: function(step) {
+            this.step = step;
+            this.render();
         },
         connect: function() {
+            this.error = null;
+            this.selectStep(Steps.SCAN_QR_CODE);
             this.clearQR();
+            if (this.timeout) {
+                clearTimeout(this.timeout);
+                this.timeout = null;
+            }
+
             var accountManager = getAccountManager();
+
             accountManager.registerSecondDevice(
                 this.setProvisioningUrl.bind(this),
-                this.confirmNumber.bind(this),
-                this.incrementCounter.bind(this)
+                this.confirmNumber.bind(this)
             ).catch(this.handleDisconnect.bind(this));
         },
         handleDisconnect: function(e) {
-            if (this.canceled) {
-                return;
-            }
             console.log('provisioning failed', e.stack);
 
+            this.error = e;
+            this.render();
+
             if (e.message === 'websocket closed') {
-                this.showConnectionError();
                 this.trigger('disconnected');
-            } else if (e.name === 'HTTPError' && e.code == -1) {
-                this.selectStep(Steps.NETWORK_ERROR);
-            } else if (e.name === 'HTTPError' && e.code == 411) {
-                this.showTooManyDevices();
-            } else {
+            } else if (e.name !== 'HTTPError' || (e.code !== -1 && e.code !== 411)) {
```

Comment:
Could we give `-1` and `411` names for clarity to avoid having magic numbers?

Reply:
Yup!

Reply:
🙇